File: codon/parser/visitors/typecheck/collections.cpp
// Copyright (C) 2022-2025 Exaloop Inc. <https://exaloop.io>

#include "codon/parser/ast.h"
#include "codon/parser/cache.h"
#include "codon/parser/common.h"
#include "codon/parser/visitors/typecheck/typecheck.h"

using fmt::format;
using namespace codon::error;

namespace codon::ast {

using namespace types;

/// Transform tuples.
/// @example
///   `(a1, ..., aN)` -> `Tuple.__new__(a1, ..., aN)`
void TypecheckVisitor::visit(TupleExpr *expr) {
  resultExpr =
      transform(N<CallExpr>(N<DotExpr>(N<IdExpr>(TYPE_TUPLE), "__new__"), expr->items));
}

/// Transform a list `[a1, ..., aN]` to the corresponding statement expression.
/// See @c transformComprehension
void TypecheckVisitor::visit(ListExpr *expr) {
  expr->setType(instantiateUnbound());
  auto name = getStdLibType("List")->name;
  if ((resultExpr = transformComprehension(name, "append", expr->items))) {
    resultExpr->setAttribute(Attr::ExprList);
  }
}

/// Transform a set `{a1, ..., aN}` to the corresponding statement expression.
/// See @c transformComprehension
void TypecheckVisitor::visit(SetExpr *expr) {
  expr->setType(instantiateUnbound());
  auto name = getStdLibType("Set")->name;
  if ((resultExpr = transformComprehension(name, "add", expr->items))) {
    resultExpr->setAttribute(Attr::ExprSet);
  }
}

/// Transform a dictionary `{k1: v1, ..., kN: vN}` to a corresponding statement
/// expression. See @c transformComprehension
void TypecheckVisitor::visit(DictExpr *expr) {
  expr->setType(instantiateUnbound());
  auto name = getStdLibType("Dict")->name;
  if ((resultExpr = transformComprehension(name, "__setitem__", expr->items))) {
    resultExpr->setAttribute(Attr::ExprDict);
  }
}

/// Transform a tuple generator expression.
/// @example
///   `tuple(expr for i in tuple_generator)` -> `Tuple.N.__new__(expr...)`
void TypecheckVisitor::visit(GeneratorExpr *expr) {
  // List comprehension optimization:
  // Use `iter.__len__()` when creating list if there is a single for loop
  // without any if conditions in the comprehension
  bool canOptimize =
      expr->kind == GeneratorExpr::ListGenerator && expr->loopCount() == 1;
  if (canOptimize) {
    auto iter = transform(clone(cast<ForStmt>(expr->getFinalSuite())->getIter()));
    auto ce = cast<CallExpr>(iter);
    IdExpr *id = nullptr;
    if (ce && (id = cast<IdExpr>(ce->getExpr()))) {
      // Turn off this optimization for static items
      canOptimize &=
          !startswith(id->getValue(), "std.internal.types.range.staticrange");
      canOptimize &= !startswith(id->getValue(), "statictuple");
    }
  }

  Expr *var = N<IdExpr>(getTemporaryVar("gen"));
  if (expr->kind == GeneratorExpr::ListGenerator) {
    // List comprehensions
    expr->setFinalExpr(
        N<CallExpr>(N<DotExpr>(clone(var), "append"), expr->getFinalExpr()));
    auto suite = expr->getFinalSuite();
    auto noOptStmt =
        N<SuiteStmt>(N<AssignStmt>(clone(var), N<CallExpr>(N<IdExpr>("List"))), suite);

    if (canOptimize) {
      auto optimizeVar = getTemporaryVar("i");
      auto origIter = cast<ForStmt>(expr->getFinalSuite())->getIter();

      auto optStmt = clone(noOptStmt);
      cast<ForStmt>((*cast<SuiteStmt>(optStmt))[1])->iter = N<IdExpr>(optimizeVar);
      optStmt = N<SuiteStmt>(
          N<AssignStmt>(N<IdExpr>(optimizeVar), clone(origIter)),
          N<AssignStmt>(
              clone(var),
              N<CallExpr>(N<IdExpr>("List"),
                          N<CallExpr>(N<DotExpr>(N<IdExpr>(optimizeVar), "__len__")))),
          (*cast<SuiteStmt>(optStmt))[1]);
      resultExpr = N<IfExpr>(
          N<CallExpr>(N<IdExpr>("hasattr"), clone(origIter), N<StringExpr>("__len__")),
          N<StmtExpr>(optStmt, clone(var)), N<StmtExpr>(noOptStmt, var));
    } else {
      resultExpr = N<StmtExpr>(noOptStmt, var);
    }
    resultExpr = transform(resultExpr);
  } else if (expr->kind == GeneratorExpr::SetGenerator) {
    // Set comprehensions
    auto head = N<AssignStmt>(clone(var), N<CallExpr>(N<IdExpr>("Set")));
    expr->setFinalExpr(
        N<CallExpr>(N<DotExpr>(clone(var), "add"), expr->getFinalExpr()));
    auto suite = expr->getFinalSuite();
    resultExpr = transform(N<StmtExpr>(N<SuiteStmt>(head, suite), var));
  } else if (expr->kind == GeneratorExpr::DictGenerator) {
    // Dictionary comprehensions
    auto head = N<AssignStmt>(clone(var), N<CallExpr>(N<IdExpr>("Dict")));
    expr->setFinalExpr(N<CallExpr>(N<DotExpr>(clone(var), "__setitem__"),
                                   N<StarExpr>(expr->getFinalExpr())));
    auto suite = expr->getFinalSuite();
    resultExpr = transform(N<StmtExpr>(N<SuiteStmt>(head, suite), var));
  } else if (expr->kind == GeneratorExpr::TupleGenerator) {
    seqassert(expr->loopCount() == 1, "invalid tuple generator");
    auto gen = transform(cast<ForStmt>(expr->getFinalSuite())->getIter());
    if (!gen->getType()->canRealize())
      return; // Wait until the iterator can be realized

    auto block = N<SuiteStmt>();
    // `tuple = tuple_generator`
    auto tupleVar = getTemporaryVar("tuple");
    block->addStmt(N<AssignStmt>(N<IdExpr>(tupleVar), gen));

    auto forStmt = clone(cast<ForStmt>(expr->getFinalSuite()));
    auto finalExpr = expr->getFinalExpr();
    auto [ok, delay, preamble, staticItems] = transformStaticLoopCall(
        cast<ForStmt>(expr->getFinalSuite())->getVar(), &forStmt->suite, gen,
        [&](Stmt *wrap) { return N<StmtExpr>(clone(wrap), clone(finalExpr)); }, true);
    if (!ok)
      E(Error::CALL_BAD_ITER, gen, gen->getType()->prettyString());
    if (delay)
      return;

    std::vector<Expr *> tupleItems;
    for (auto &i : staticItems)
      tupleItems.push_back(cast<Expr>(i));
    if (preamble)
      block->addStmt(preamble);
    resultExpr = transform(N<StmtExpr>(block, N<TupleExpr>(tupleItems)));
  } else {
    expr->loops =
        transform(expr->getFinalSuite()); // assume: internal data will be changed
    unify(expr->getType(), instantiateType(getStdLibType("Generator"),
                                           {expr->getFinalExpr()->getType()}));
    if (realize(expr->getType()))
      expr->setDone();
  }
}

/// Transform a collection of type `type` to a statement expression:
///   `[a1, ..., aN]` -> `cont = [type](); (cont.[fn](a1); ...); cont`
/// Any star-expression within the collection will be expanded:
///   `[a, *b]` -> `cont.[fn](a); for i in b: cont.[fn](i)`.
/// @example
///   `[a, *b, c]`  -> ```cont = List(3)
///                       cont.append(a)
///                       for i in b: cont.append(i)
///                       cont.append(c)```
///   `{a, *b, c}`  -> ```cont = Set()
///                       cont.add(a)
///                       for i in b: cont.add(i)
///                       cont.add(c)```
///   `{a: 1, **d}` -> ```cont = Dict()
///                       cont.__setitem__((a, 1))
///                       for i in b.items(): cont.__setitem__((i[0], i[i]))```
Expr *TypecheckVisitor::transformComprehension(const std::string &type,
                                               const std::string &fn,
                                               std::vector<Expr *> &items) {
  // Deduce the super type of the collection--- in other words, the least common
  // ancestor of all types in the collection. For example, `type([1, 1.2]) == type([1.2,
  // 1]) == float` because float is an "ancestor" of int.
  auto superTyp = [&](ClassType *collectionCls, ClassType *ti) -> TypePtr {
    if (!collectionCls)
      return ti->shared_from_this();
    if (collectionCls->is("int") && ti->is("float")) {
      // Rule: int derives from float
      return ti->shared_from_this();
    } else if (collectionCls->name != TYPE_OPTIONAL && ti->name == TYPE_OPTIONAL) {
      // Rule: T derives from Optional[T]
      return instantiateType(getStdLibType("Optional"),
                             std::vector<types::Type *>{collectionCls});
    } else if (collectionCls->name == TYPE_OPTIONAL && ti->name != TYPE_OPTIONAL) {
      return instantiateType(getStdLibType("Optional"), std::vector<types::Type *>{ti});
    } else if (!collectionCls->is("pyobj") && ti->is("pyobj")) {
      // Rule: anything derives from pyobj
      return ti->shared_from_this();
    } else if (collectionCls->name != ti->name) {
      // Rule: subclass derives from superclass
      const auto &mros = getClass(collectionCls)->mro;
      for (size_t i = 1; i < mros.size(); i++) {
        auto t = instantiateType(mros[i].get(), collectionCls);
        if (t->unify(ti, nullptr) >= 0) {
          return ti->shared_from_this();
          break;
        }
      }
    }
    return nullptr;
  };

  TypePtr collectionTyp = instantiateUnbound();
  bool done = true;
  bool isDict = type == getStdLibType("Dict")->name;
  for (auto &i : items) {
    ClassType *typ = nullptr;
    if (!isDict && cast<StarExpr>(i)) {
      auto star = cast<StarExpr>(i);
      star->expr = transform(N<CallExpr>(N<DotExpr>(star->getExpr(), "__iter__")));
      if (star->getExpr()->getType()->is("Generator"))
        typ = extractClassGeneric(star->getExpr()->getType())->getClass();
    } else if (isDict && cast<KeywordStarExpr>(i)) {
      auto star = cast<KeywordStarExpr>(i);
      star->expr = transform(N<CallExpr>(N<DotExpr>(star->getExpr(), "items")));
      if (star->getExpr()->getType()->is("Generator"))
        typ = extractClassGeneric(star->getExpr()->getType())->getClass();
    } else {
      i = transform(i);
      typ = i->getClassType();
    }
    if (!typ) {
      done = false;
      continue;
    }
    if (!collectionTyp->getClass()) {
      unify(collectionTyp.get(), typ);
    } else if (!isDict) {
      if (auto t = superTyp(collectionTyp->getClass(), typ))
        collectionTyp = t;
    } else {
      auto tt = unify(typ, instantiateType(generateTuple(2)))->getClass();
      seqassert(collectionTyp->getClass() &&
                    collectionTyp->getClass()->generics.size() == 2 &&
                    tt->generics.size() == 2,
                "bad dict");
      std::vector<types::TypePtr> nt;
      for (int di = 0; di < 2; di++) {
        nt.push_back(extractClassGeneric(collectionTyp.get(), di)->shared_from_this());
        if (!nt[di]->getClass())
          unify(nt[di].get(), extractClassGeneric(tt, di));
        else if (auto dt = superTyp(nt[di]->getClass(),
                                    extractClassGeneric(tt, di)->getClass()))
          nt[di] = dt;
      }
      collectionTyp =
          instantiateType(generateTuple(nt.size()), ctx->cache->castVectorPtr(nt));
    }
  }
  if (!done)
    return nullptr;
  std::vector<Stmt *> stmts;
  Expr *var = N<IdExpr>(getTemporaryVar("cont"));

  std::vector<Expr *> constructorArgs{};
  if (type == getStdLibType("List")->name && !items.empty()) {
    // Optimization: pre-allocate the list with the exact number of elements
    constructorArgs.push_back(N<IntExpr>(items.size()));
  }
  auto t = N<IdExpr>(type);
  auto ta = instantiateType(getStdLibType(type));
  if (isDict && collectionTyp->getClass()) {
    seqassert(collectionTyp->getClass()->isRecord(), "bad dict");
    std::vector<types::Type *> nt;
    for (auto &g : collectionTyp->getClass()->generics)
      nt.push_back(g.getType());
    ta = instantiateType(getStdLibType(type), nt);
  } else if (!isDict) {
    ta = instantiateType(getStdLibType(type), {collectionTyp.get()});
  }
  t->setType(instantiateTypeVar(ta.get()));
  stmts.push_back(N<AssignStmt>(clone(var), N<CallExpr>(t, constructorArgs)));
  for (const auto &it : items) {
    if (!isDict && cast<StarExpr>(it)) {
      // Unpack star-expression by iterating over it
      // `*star` -> `for i in star: cont.[fn](i)`
      auto star = cast<StarExpr>(it);
      Expr *forVar = N<IdExpr>(getTemporaryVar("i"));
      star->getExpr()->setAttribute(Attr::ExprStarSequenceItem);
      stmts.push_back(N<ForStmt>(
          clone(forVar), star->getExpr(),
          N<ExprStmt>(N<CallExpr>(N<DotExpr>(clone(var), fn), clone(forVar)))));
    } else if (isDict && cast<KeywordStarExpr>(it)) {
      // Expand kwstar-expression by iterating over it: see the example above
      auto star = cast<KeywordStarExpr>(it);
      Expr *forVar = N<IdExpr>(getTemporaryVar("it"));
      star->getExpr()->setAttribute(Attr::ExprStarSequenceItem);
      stmts.push_back(N<ForStmt>(
          clone(forVar), star->getExpr(),
          N<ExprStmt>(N<CallExpr>(N<DotExpr>(clone(var), fn),
                                  N<IndexExpr>(clone(forVar), N<IntExpr>(0)),
                                  N<IndexExpr>(clone(forVar), N<IntExpr>(1))))));
    } else {
      it->setAttribute(Attr::ExprSequenceItem);
      if (isDict) {
        stmts.push_back(N<ExprStmt>(N<CallExpr>(N<DotExpr>(clone(var), fn),
                                                N<IndexExpr>(it, N<IntExpr>(0)),
                                                N<IndexExpr>(it, N<IntExpr>(1)))));
      } else {
        stmts.push_back(N<ExprStmt>(N<CallExpr>(N<DotExpr>(clone(var), fn), it)));
      }
    }
  }
  return transform(N<StmtExpr>(stmts, var));
}

} // namespace codon::ast

File: meister/converted/passes/typecheck/collections.py
# Copyright (C) 2022-2026 Exaloop Inc. <https://exaloop.io>

from __future__ import annotations

from typing import TYPE_CHECKING

from ....bridge import List, cast
from ... import ast
from . import classes, infer, loops, utils
from .ctx import TypecheckError

if TYPE_CHECKING:
    from . import TypeVisitor


def typecheck_tuple(self: TypeVisitor, node: ast.TupleExpr) -> ast.Node:
    """
    Transform tuples.
    @example
    `(a1, ..., aN)` -> `Tuple.__new__(a1, ..., aN)`
    """

    result = self.visit(
        ast.CallExpr(
            ast.DotExpr(ast.IdExpr(ast.types.Stdlib.Tuple), member="__new__"),
            items=node.items,
        )
    )
    return result


def typecheck_list(self: TypeVisitor, node: ast.ListExpr) -> ast.Node:
    """
    Transform a list `[a1, ..., aN]` to the corresponding statement expression.
    See @c transformComprehension
    """

    node.type = utils.instantiate_unbound(self.ctx)
    result = transform_comprehension(self, ast.types.Stdlib.List, node.items, method="append")
    if result:
        result.set(ast.Attr.ExprList)
        return result
    return node


def typecheck_set(self: TypeVisitor, node: ast.SetExpr) -> ast.Node:
    """
    Transform a set `{a1, ..., aN}` to the corresponding statement expression.
    See @c transformComprehension
    """

    node.type = utils.instantiate_unbound(self.ctx)
    result = transform_comprehension(self, ast.types.Stdlib.Set, node.items, method="add")
    if result:
        result.set(ast.Attr.ExprSet)
        return result
    return node


def typecheck_dict(self: TypeVisitor, node: ast.DictExpr) -> ast.Node:
    """
    Transform a dictionary `{k1: v1, ..., kN: vN}` to a corresponding statement
    expression. See @c transformComprehension
    """

    node.type = utils.instantiate_unbound(self.ctx)
    result = transform_comprehension(self, ast.types.Stdlib.Dict, node.items, method="__setitem__")
    if result:
        result.set(ast.Attr.ExprDict)
        return result
    return node


def typecheck_generator(self: TypeVisitor, node: ast.GeneratorExpr) -> ast.Node:
    """
    Transform a tuple generator expression.
    @example
    `tuple(expr for i in tuple_generator)` -> `Tuple.N.__new__(expr...)`
    """

    # List comprehension optimization:
    # Use `iter.__len__()` when creating list if there is a single for loop
    # without any if conditions in the comprehension
    optimize = node.kind is ast.GeneratorExpr.Kind.ListGenerator and node.loop_count() == 1
    final = node.final_suite()
    if optimize:
        # Turn off this optimization for static items
        match self.visit(final.iter.clone()):
            case ast.CallExpr(ast.IdExpr(value=name)) if not name.startswith("std.internal.static"):
                optimize = False
    var = ast.IdExpr(utils.get_temporary_var(self.ctx, "gen"))
    expr = node.final_expr()

    if node.kind is ast.GeneratorExpr.Kind.ListGenerator:
        # List comprehensions
        node.set_final_expr(ast.CallExpr(ast.DotExpr(var.clone(), member="append"), items=[expr]))
        plain = ast.SuiteStmt(
            ast.AssignStmt(var.clone(), rhs=ast.CallExpr(ast.IdExpr(ast.types.Stdlib.List))),
            node.loops,
        )
        if optimize:
            opt_var = utils.get_temporary_var(self.ctx, "i")
            opt_for = node.loops.clone()
            opt_for.iter = ast.IdExpr(opt_var)
            opt = ast.SuiteStmt(
                ast.AssignStmt(ast.IdExpr(opt_var), rhs=final.iter.clone()),
                ast.AssignStmt(
                    var.clone(),
                    rhs=ast.CallExpr(
                        ast.IdExpr(ast.types.Stdlib.List),
                        items=[ast.CallExpr(ast.DotExpr(ast.IdExpr(opt_var), member="__len__"))],
                    ),
                ),
                opt_for,
            )
            result = ast.IfExpr(
                ast.CallExpr(
                    ast.IdExpr("hasattr"), items=[final.iter.clone(), ast.StringExpr("__len__")]
                ),
                ifexpr=ast.StmtExpr(opt.items, expr=var.clone()),
                elsexpr=ast.StmtExpr(plain.items, expr=var),
            )
        else:
            result = ast.StmtExpr(plain.items, expr=var)
        return self.visit(result)

    if node.kind is ast.GeneratorExpr.Kind.SetGenerator:
        # Set comprehensions
        head = ast.AssignStmt(var.clone(), rhs=ast.CallExpr(ast.IdExpr(ast.types.Stdlib.Set)))
        node.set_final_expr(ast.CallExpr(ast.DotExpr(var.clone(), member="add"), items=[expr]))
        return self.visit(ast.StmtExpr([head, node.loops], expr=var))
    elif node.kind is ast.GeneratorExpr.Kind.DictGenerator:
        # Dictionary comprehensions
        head = ast.AssignStmt(var.clone(), rhs=ast.CallExpr(ast.IdExpr(ast.types.Stdlib.Dict)))
        node.set_final_expr(
            ast.CallExpr(ast.DotExpr(var.clone(), member="__setitem__"), items=[ast.StarExpr(expr)])
        )
        return self.visit(ast.StmtExpr([head, node.loops], expr=var))
    elif node.kind is ast.GeneratorExpr.Kind.TupleGenerator:
        assert node.loop_count() == 1, "invalid tuple generator"
        generator_node = self.visit(final.iter)
        if not generator_node.type.can_realize():
            return node  # Wait until the iterator can be realized

        # `tuple = tuple_generator`
        tuple_name = utils.get_temporary_var(self.ctx, "tuple")
        block = ast.SuiteStmt(ast.AssignStmt(ast.IdExpr(tuple_name), rhs=generator_node))
        static_items = loops.populate_static_loop(
            self,
            final.var,
            final.suite,
            generator_node,
            expr,
        )

        return self.visit(ast.StmtExpr(block.items, expr=ast.TupleExpr(static_items)))
    else:
        node.loops = self.visit(node.loops)  # assume: internal data will be changed
        expr = node.final_expr()
        if not expr:
            # Case such as (0 for _ in static.range(2))
            # TODO: make this better.
            raise TypecheckError(
                node,
                "generator cannot be compiled. If using static tuple generator, use "
                "tuple(...) instead.",
            )
        infer.unify(
            node.type,
            utils.instantiate_type(
                self.ctx,
                utils.get_stdlib_type(self.ctx, ast.types.Stdlib.Generator),
                [expr.type],
            ),
        )
        if infer.realize(node.type):
            node.done = True
        return node


def transform_comprehension(
    self: TypeVisitor,
    type_name: str,
    items: List[ast.Expr],
    method: str,
) -> ast.Expr | None:
    """
    Transform a collection of type `type` to a statement expression:
    `[a1, ..., aN]` -> `cont = [type](); (cont.[fn](a1); ...); cont`
    Any star-expression within the collection will be expanded:
    `[a, *b]` -> `cont.[fn](a); for i in b: cont.[fn](i)`.
    @example
    `[a, *b, c]`  -> ```cont = List(3)
    cont.append(a)
    for i in b: cont.append(i)
    cont.append(c)```
    `{a, *b, c}`  -> ```cont = Set()
    cont.add(a)
    for i in b: cont.add(i)
    cont.add(c)```
    `{a: 1, **d}` -> ```cont = Dict()
    cont.__setitem__((a, 1))
    for i in b.items(): cont.__setitem__((i[0], i[i]))```
    """

    def lowest_common_type(
        typ: ast.types.Class | None, item_type: ast.types.Class
    ) -> ast.types.Type | None:
        if not typ:
            return item_type
        elif typ.is_type("int") and item_type.is_type("float"):
            return item_type
        elif typ.name != ast.types.Stdlib.Optional and item_type.name == ast.types.Stdlib.Optional:
            return utils.instantiate_type(
                self.ctx,
                utils.get_stdlib_type(self.ctx, ast.types.Stdlib.Optional),
                [typ],
            )
        elif typ.name == ast.types.Stdlib.Optional and item_type.name != ast.types.Stdlib.Optional:
            return utils.instantiate_type(
                self.ctx,
                utils.get_stdlib_type(self.ctx, ast.types.Stdlib.Optional),
                [item_type],
            )
        elif not typ.is_type("pyobj") and item_type.is_type("pyobj"):
            return item_type
        elif typ.name != item_type.name:
            cls_data = utils.get_class(self.ctx, typ)
            item_cls_data = utils.get_class(self.ctx, item_type)
            if cls_data and item_cls_data:
                for collection_mro in cls_data.mro:
                    typ = utils.instantiate_type(
                        self.ctx, collection_mro, [g.type for g in typ.generics]
                    )
                    for item_mro in item_cls_data.mro:
                        candidate = utils.instantiate_type(
                            self.ctx, item_mro, [g.type for g in item_type.generics]
                        )
                        if typ.unify(candidate) >= 0:
                            return typ
        return None

    collection_type = utils.instantiate_unbound(self.ctx)
    done = True
    is_dict = type_name == ast.types.Stdlib.Dict
    for idx, item in enumerate(items):
        # Deduce the lowest common type of the collection--- in other words, the lowest
        # common ancestor of all types in the collection. For example, `type([1, 1.2]) ==
        # type([1.2, 1]) == float` because float is an "ancestor" of int.
        # TODO: use wrapExpr...

        item_type = None
        if not is_dict and isinstance(item, ast.StarExpr):
            item.expr = self.visit(ast.CallExpr(ast.DotExpr(item.expr, member="__iter__")))
            if item.expr and item.expr.type and item.expr.type.is_type("Generator"):
                item_type = item.expr.type[0]
        elif is_dict and isinstance(item, ast.KeywordStarExpr):
            item.expr = self.visit(ast.CallExpr(ast.DotExpr(item.expr, member="items")))
            if item.expr and item.expr.type and item.expr.type.is_type("Generator"):
                item_type = item.expr.type[0]
        else:
            items[idx] = item = self.visit(item)
            item_type = item.get_class_type()
        if not item_type:
            done = False
            continue

        if not collection_type:
            infer.unify(collection_type, item_type)
        elif not is_dict:
            if common := lowest_common_type(collection_type, item_type):
                collection_type = common
        else:
            tuple_type = infer.unify(
                item_type,
                utils.instantiate_type(self.ctx, classes.generate_tuple(self, 2)),
            )
            assert collection_type.is_record() and len(collection_type.generics) == 2
            assert len(tuple_type.generics) == 2

            new_types = []
            for dict_index in range(2):
                new_type = collection_type[dict_index]
                if not new_type:
                    infer.unify(new_type, tuple_type[dict_index])
                elif common := lowest_common_type(new_type, tuple_type[dict_index]):
                    new_type = common
                new_types.append(new_type)
            collection_type = utils.instantiate_type(
                self.ctx, classes.generate_tuple(self, len(new_types)), new_types
            )
    if not done:
        return None

    stmts = []
    var = ast.IdExpr(utils.get_temporary_var(self.ctx, "cont"))
    ctr_args: List[ast.CallExpr.Arg] = []
    if type_name == ast.types.Stdlib.List and items:
        ctr_args.append(ast.IntExpr(len(items)))
    ctr = ast.IdExpr(type_name)
    collection_template = utils.instantiate_type(
        self.ctx, utils.get_stdlib_type(self.ctx, type_name)
    )
    if is_dict and collection_type:
        assert collection_type.is_record()
        collection_template = utils.instantiate_type(
            self.ctx,
            utils.get_stdlib_type(self.ctx, type_name),
            [generic.type for generic in collection_type.generics],
        )
    elif not is_dict:
        collection_template = utils.instantiate_type(
            self.ctx,
            utils.get_stdlib_type(self.ctx, type_name),
            [collection_type],
        )
    ctr.type = utils.instantiate_type_var(self.ctx, collection_template)
    stmts.append(ast.AssignStmt(var.clone(), rhs=ast.CallExpr(ctr, items=ctr_args)))
    for item in items:
        if not is_dict and isinstance(item, ast.StarExpr):
            # `*star` -> `for i in star: cont.[fn](i)`
            loop_var = ast.IdExpr(utils.get_temporary_var(self.ctx, "i"))
            item.expr.set(ast.Attr.ExprStarSequenceItem)
            stmts.append(
                ast.ForStmt(
                    loop_var.clone(),
                    iter=item.expr,
                    suite=ast.ExprStmt(
                        ast.CallExpr(
                            ast.DotExpr(var.clone(), member=method), items=[loop_var.clone()]
                        )
                    ),
                )
            )
        elif is_dict and isinstance(item, ast.KeywordStarExpr):
            # Same for **kwstars
            loop_var = ast.IdExpr(utils.get_temporary_var(self.ctx, "it"))
            item.expr.set(ast.Attr.ExprStarSequenceItem)
            args = [
                ast.IndexExpr(loop_var.clone(), index=ast.IntExpr(0)),
                ast.IndexExpr(loop_var.clone(), index=ast.IntExpr(1)),
            ]
            stmts.append(
                ast.ForStmt(
                    loop_var.clone(),
                    iter=item.expr,
                    suite=ast.ExprStmt(
                        ast.CallExpr(ast.DotExpr(var.clone(), member=method), items=args)
                    ),
                )
            )
        else:
            item.set(ast.Attr.ExprSequenceItem)
            args = []
            if is_dict:
                head = item
                if utils.has_side_effect(head):
                    temporary = utils.get_temporary_var(self.ctx, "star")
                    lead = ast.AssignExpr(ast.IdExpr(temporary), expr=head)
                    head = ast.IdExpr(temporary)
                else:
                    lead = head.clone()
                lead.set(ast.Attr.ExprSequenceItem)
                head.set(ast.Attr.ExprSequenceItem)
                args = [
                    ast.IndexExpr(lead, index=ast.IntExpr(0)),
                    ast.IndexExpr(head, index=ast.IntExpr(1)),
                ]
            else:
                args = [item]
            stmts.append(
                ast.ExprStmt(ast.CallExpr(ast.DotExpr(var.clone(), member=method), items=args))
            )
    return self.visit(ast.StmtExpr(stmts, expr=var))
